scraping_utils.py

```diff
@@ -293,4 +293,43 @@ def extract_timetables(string, year, term):
         out["courses"].append(course_out)
     return out
 
+def distribute_shared_sections(out):
+    messages = []
+    # For each course
+    for index, course in enumerate(out["courses"]):
+        course_code = course["subject_code"] + course["course_number"]
+        sec_comps = {}
+        for section in course["sections"]:
+            sec_comps[section["id"]] = []
+            for component in section["components"]:
+                if component["type"] not in sec_comps[section["id"]]:
+                    sec_comps[section["id"]].append(component["type"])
 
+        arr_lookup = lambda elt: len(sec_comps[elt["id"]])
+        comp_secs = group_by_eq(course["sections"], arr_lookup)
+
+        if len(comp_secs) >= 1:
+            continue
+
+        sections = []
+        bad_sec_ids = []
+        for section in course["sections"]:
+            if section["id"] in bad_sec_ids:
+                continue
+            for bad_section in comp_secs[1]:
+                if (len(sec_comps[section["id"]]) == 1) and (sec_comps[bad_section["id"]][0] == sec_comps[section["id"]][0]):
+                    continue
+                section["components"] = section["components"] + bad_section["components"]
+
+                bad_sec_ids[bad_section["id"]] = 0
```

Comment:
```suggestion
                bad_sec_ids.add(bad_section["id"])
```
Since all this is used for is to prevent distributed sections from being distributed to, I think a set would be a bit more efficient (even if just in principle, as these are all very small samples).  Since python arrays are typically indexed by integers, I'm actually a little bit surprised this worked.  Was there some automatic type casting going on?
E.g.
```
In [7]: l = []

In [8]: l["d"] = 0
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-bdb2604f7990> in <module>()
----> 1 l["d"] = 0

TypeError: list indices must be integers or slices, not str
```

Reply:
Wow I'm honestly surprised this worked at all O.o ! I guess the testing I did didn't have bad sections so it didn't matter. I think the next thing I'll tackle is some unit testing so mistakes like this are caught way sooner, I'll wait on integration tests since we're not sure exactly what all the API will look like but at least some unit tests for what we have would be good.